### PR TITLE
Provide actionable error message when validation fails

### PR DESCRIPTION
**Pull request by geoand**
This plugin is used extensively in Quarkus and new contributors
are often left unsure of what to do when they see the validation
error message.
This change provides users with an explicit command to run when the
validation fails

### CODE REVIEW

File: src/main/java/net/revelc/code/formatter/ValidateMojo.java
Comment (famod) [inline src/main/java/net/revelc/code/formatter/ValidateMojo.java:93]: `-f` would be quicker, especially for large projects like Quarkus. `MavenProject.getFile()` (or `.getBasedir()`) should give you what you'd need for that.

**Review by famod — COMMENTED** [review]
Cool!

I left some remarks and I have two more:
- for parallel builds you might end up with multiple errors, but in that case the user has to invoke multiple commands (just the way the cookie crumbles)
- this doesn't cover profiles and system properties (`-D...`), which might lead to inconsistencies

**Comment by famod** [comment]
W.r.t.
> - this doesn't cover profiles and system properties (`-D...`), which might lead to inconsistencies

I think this (as is) covers the most use cases. The message also says
> *for example* by invoking

So all in all I think this is something that can be improved later, if needed.

File: src/main/java/net/revelc/code/formatter/ValidateMojo.java
Comment (geoand) [reply src/main/java/net/revelc/code/formatter/ValidateMojo.java:88]: > Though, a full variant (with all coordinates) also covers rare cases in which the plugin isn't configured in pom.xml.

Yeah, that's what I had in mind.

> MojoExecution should provide that id.

Checking, thanks

File: src/main/java/net/revelc/code/formatter/ValidateMojo.java
Comment (geoand) [reply src/main/java/net/revelc/code/formatter/ValidateMojo.java:92]: Good catch!

**Comment by geoand** [comment]
@famod PR updated per your suggestions

**Comment by famod** [comment]
The command is probably rather long now, no? I might make sense to put it on another line (e.g. with `\n\t`), but let's wait for some feedback by the actual maintainer! 😉

File: src/main/java/net/revelc/code/formatter/ValidateMojo.java
Comment (ctubbsii) [inline src/main/java/net/revelc/code/formatter/ValidateMojo.java:92]: I don't think you'd want to include the execution ID here. What purpose does that serve?

File: src/main/java/net/revelc/code/formatter/ValidateMojo.java
Comment (ctubbsii) [inline src/main/java/net/revelc/code/formatter/ValidateMojo.java:98]: This would place the format goal before the `-f basedir/pom.xml`. Although this may work, `mvn --help` says, options, like `-f basedir/pom.xml` should be placed before the goals/phases. It'd be good to follow that consistently.

```suggestion
        return String.format("mvn%s %s", isMultiModule ? " -f " + mavenProject.getBasedir() : "", validateInvocation);
```

(This suggestion assumes 'mvn ' is removed from the front of `validateInvocation`.)

**Review by ctubbsii — CHANGES_REQUESTED** [review]

File: src/main/java/net/revelc/code/formatter/ValidateMojo.java
Comment (geoand) [reply src/main/java/net/revelc/code/formatter/ValidateMojo.java:92]: As @famod mentions above, it's used to ensure that setups that contain multiple executions of the plugin will have the proper ID.

File: src/main/java/net/revelc/code/formatter/ValidateMojo.java
Comment (geoand) [reply src/main/java/net/revelc/code/formatter/ValidateMojo.java]: Applied

File: src/main/java/net/revelc/code/formatter/ValidateMojo.java
Comment (ctubbsii) [reply src/main/java/net/revelc/code/formatter/ValidateMojo.java:92]: My thinking was that this would only execute using the config that was set up for verification. I'm not sure that makes sense. What if they have a config specifically for formatting? Although executing with the same config that is causing validate to fail is more likely to get the validation to succeed, it may not be the best command to execute. It also wouldn't work if the execution ID is defined inside a profile, unless the user also activated that profile as well.

Since we're merely making a suggestion, it's probably better to just omit the execution ID, and leave it up to the user. That way, we don't risk giving bad advice.

Also, since presumably the version is already specified inside this plugin's POM, we can probably omit the version as well, to streamline the suggestion.

Comment (ctubbsii) [reply src/main/java/net/revelc/code/formatter/ValidateMojo.java:92]: To be clear, this isn't really an objection. Just discussing. If others think it's good to have the version and the execution ID, I'm fine with it.

Comment (famod) [reply src/main/java/net/revelc/code/formatter/ValidateMojo.java:92]: In the end it's your choice @ctubbsii but when specifying the execution id you will always end up with the properly merged config of all "sources", e.g. including profiles and plugin-wide config.
AFAICS this also includes the execution id. What the plugin sees via `MojoExecution` should always be the "effective model".

Comment (ctubbsii) [reply src/main/java/net/revelc/code/formatter/ValidateMojo.java:92]: The execution ID can be reused by multiple execution definitions, resulting in effective models in the end.

The most common is the how users often clobber the default execution ID of a plugin that is bound to the default build lifecycle to change its behavior or to prevent it from running (by giving it a fake phase), as in https://stackoverflow.com/questions/34590452/what-is-the-implicit-id-of-a-maven-plugin-execution

You can also clobber an execution ID and change the effective model for that execution by reusing an execution ID from an execution defined in a parent POM.

But, the one I'm thinking about is when an execution ID is reused inside a profile, in order to conditionally clobber an execution defined outside the profile (or in a different profile). In order to get the same effective model for the suggested format command, you'd also have to activate the same set of profiles that were active when the verification failed.

For example, if you run `mvn -Prelease formatter:validate` and it fails, these changes might suggest running `mvn net.revelc.code.formatter:formatter-maven-plugin:2.17.1:format@validateReleaseFormatting`, even though that won't work without also adding `-Prelease` if the formatter execution ID was defined inside the release profile.

Comment (famod) [reply src/main/java/net/revelc/code/formatter/ValidateMojo.java:92]: > you'd also have to activate the same set of profiles that were active when the verification failed

Yeah, totally. That's what I meant earlier today, but it makes the command even longer.

I think the bottom line for now is to keep it simple(r) and see how it goes.
I'm fine with dropping the execution id.

Comment (geoand) [reply src/main/java/net/revelc/code/formatter/ValidateMojo.java]: Okay, removed

**Pull request merged**
